components/style/properties_and_values.rs

```diff
@@ -0,0 +1,1167 @@
+            a == b
+        }
+
+        #[cfg(feature = "gecko")]
+        fn idents_eq(a: &Atom, b: &str) -> bool {
+            *a == b.into()
+        }
+
+        macro_rules! parse {
+            ($_self:expr, $context:expr, $input:expr,
+
+             $($typ:ident => $fn:path),*) => {
+                match $_self {
+                    $(
+                        Type::$typ => {
+                            $fn(context, input).map(|x| {
+                                SpecifiedValueItem::$typ(x)
+                            })
+                        }
+                    ), *
+
+                    // We need to actually compare SpecificIdents,
+                    // unfortunately.
+                    Type::SpecificIdent(ref x) => {
+                        $input.expect_ident_cloned().and_then(|y| {
+                            if idents_eq(&**x, &*y) {
+                                Ok(SpecifiedValueItem::SpecificIdent(x.clone()))
+                            } else {
+                                Err(BasicParseError::UnexpectedToken(Token::Ident(y)).into())
+                            }
+                        }).map_err(|e| e.into())
+                    },
+                }
+            };
+        }
+
+        fn parse_custom_ident<'i, 't>(
+            _context: &ParserContext,
+            input: &mut Parser<'i, 't>,
+        ) -> Result<Ident, StyleTraitsParseError<'i>> {
+            input.expect_ident_cloned()
+                 .map(|x| Ident((&*x).into()))
+                 .map_err(|e| e.into())
+        }
+
+        parse! {
+            *self, context, input,
+
+            Length => specified::Length::parse,
+            Number => specified::Number::parse,
+            Percentage => specified::Percentage::parse,
+            LengthPercentage => specified::LengthOrPercentage::parse,
+            Color => specified::Color::parse,
+            Image => specified::Image::parse,
+            Url => specified::url::SpecifiedUrl::parse,
+            Integer => specified::Integer::parse,
+            Angle => specified::Angle::parse,
+            Time => specified::Time::parse,
+            TransformList => transform::parse,
+            CustomIdent => parse_custom_ident
+        }
+    }
+}
+
+
+/// A custom property syntax string that is either some basic syntax string
+/// (e.g. some <url> value) or some list term. A list term syntax string allows
+/// a space-separated list of one or more repetitions of the type specified by
+/// the string. Used to build up disjunctions.
+#[derive(Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct Term {
+    /// The type of the term (e.g. <integer>).
+    pub typ: Type,
+    /// Whether or not the term is a list, i.e., if the syntax string was
+    /// <integer>+.
+    pub list: bool,
+}
+
+/// A custom property syntax string.
+#[derive(Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum Syntax {
```

Comment:
I think this could be on its own module, along with the parser, what do you think?

Also, this could land before, with a few unit tests, and such.

Maybe it's time for `custom_properties` to be a folder on its own... :)

Reply:
Hm, I am not sure I agree. This module is the same size as `custom_properties`. Also, I am not too clear on the value of unit tests that can only be run with Servo when we have web-platform tests that should be better maintained because they can be run with all web browsers.

Reply:
Sure, I meant as a stop gap, until all the extra support for exposing it lands.

Reply:
We've done that before for some CSS properties and such.